File: articles/genes-environment-and-developmental-plasticity/python/model_genes_environment_plasticity.py
#!/usr/bin/env python3
"""Model synthetic gene-environment interaction and developmental plasticity."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "genes_environment_plasticity_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_genes_environment_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["child_id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("child_id")["development_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    development_model = smf.ols(
        """
        development_score ~ lag_score + time + bio_sensitivity +
        current_care + current_stress + current_nutrition +
        school_support + neighborhood_safety + service_access +
        early_exposure + timing_weight + embedded_stress +
        embedded_support + intervention_support +
        bio_sensitivity:current_care + bio_sensitivity:current_stress
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    embedding_model = smf.ols(
        """
        embedded_stress ~ time + current_stress + current_care +
        current_nutrition + school_support + neighborhood_safety +
        service_access + bio_sensitivity + intervention_support
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_plasticity_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(development_model.summary().as_text())
        f.write("\n\nEMBEDDED STRESS MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(embedding_model.summary().as_text())

    trajectory = panel.groupby(["time", "early_exposure"], as_index=False).agg(
        average_development=("development_score", "mean"),
        standard_error=("development_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_development"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_development"] + 1.96 * trajectory["standard_error"]
    trajectory["group"] = trajectory["early_exposure"].map({0: "No early exposure", 1: "Early exposure"})
    trajectory.to_csv(OUTPUTS_DIR / "python_development_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("group"):
        plt.plot(group["time"], group["average_development"], marker="o", label=group_name)
    plt.xlabel("Time")
    plt.ylabel("Average development score")
    plt.title("Synthetic Genes, Environment, and Developmental Plasticity")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_development_trajectory.png", dpi=160)
    plt.close()

    embedded = panel.groupby("time", as_index=False).agg(
        average_embedded_stress=("embedded_stress", "mean"),
        average_embedded_support=("embedded_support", "mean"),
        average_stress=("current_stress", "mean"),
        average_care=("current_care", "mean"),
    )
    embedded.to_csv(OUTPUTS_DIR / "python_embedded_exposure_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(embedded["time"], embedded["average_embedded_stress"], marker="o", label="embedded stress")
    plt.plot(embedded["time"], embedded["average_embedded_support"], marker="o", label="embedded support")
    plt.xlabel("Time")
    plt.ylabel("Average embedded exposure")
    plt.title("Synthetic Embedded Stress and Support")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_embedded_exposure_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("sensitivity_stress_profile", as_index=False).agg(
        children=("child_id", "nunique"),
        average_development=("development_score", "mean"),
        average_embedded_stress=("embedded_stress", "mean"),
        average_embedded_support=("embedded_support", "mean"),
        average_care=("current_care", "mean"),
        average_stress=("current_stress", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_sensitivity_stress_profiles.csv", index=False)

    print("Wrote Python plasticity model outputs.")


if __name__ == "__main__":
    main()

File: articles/genes-environment-and-developmental-plasticity/r/model_genes_environment_plasticity.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "genes_environment_plasticity_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_genes_environment_panel.py first.")

panel <- read.csv(data_path)

development_model <- lmer(
  development_score ~ time + bio_sensitivity + current_care +
    current_stress + current_nutrition + school_support +
    neighborhood_safety + service_access + early_exposure +
    timing_weight + embedded_stress + embedded_support +
    intervention_support + bio_sensitivity:current_care +
    bio_sensitivity:current_stress +
    (1 + time | context_id/child_id),
  data = panel
)

embedding_model <- lmer(
  embedded_stress ~ time + current_stress + current_care +
    current_nutrition + school_support + neighborhood_safety +
    service_access + bio_sensitivity + intervention_support +
    (1 + time | context_id/child_id),
  data = panel
)

capture.output(
  list(
    development_model = summary(development_model),
    embedding_model = summary(embedding_model)
  ),
  file = file.path(outputs_dir, "r_plasticity_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, early_exposure) |>
  summarize(
    average_development = mean(development_score),
    standard_error = sd(development_score) / sqrt(n()),
    lower = average_development - 1.96 * standard_error,
    upper = average_development + 1.96 * standard_error,
    .groups = "drop"
  ) |>
  mutate(group = ifelse(early_exposure == 1, "Early exposure", "No early exposure"))

write.csv(trajectory, file.path(outputs_dir, "r_development_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_development_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_development, linetype = group)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = group), alpha = 0.12) +
    labs(title = "Synthetic Genes, Environment, and Developmental Plasticity", x = "Time", y = "Average development", linetype = "Group") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

embedded <- panel |>
  group_by(time) |>
  summarize(
    average_embedded_stress = mean(embedded_stress),
    average_embedded_support = mean(embedded_support),
    average_care = mean(current_care),
    average_stress = mean(current_stress),
    .groups = "drop"
  )

write.csv(embedded, file.path(outputs_dir, "r_embedded_exposure_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_embedded_exposure_trajectory.png"),
  plot = ggplot(embedded, aes(x = time)) +
    geom_line(aes(y = average_embedded_stress, linetype = "embedded stress"), linewidth = 1) +
    geom_line(aes(y = average_embedded_support, linetype = "embedded support"), linewidth = 1) +
    labs(title = "Synthetic Embedded Stress and Support", x = "Time", y = "Average embedded exposure", linetype = "Exposure") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R plasticity outputs.")
